src/SmartRange.ts

```diff
@@ -0,0 +1,207 @@
+/**
+ * # SmartRange
+ * offers a python like range object with "lazy" implementation
+ * @example
+ * ```ts
+ * const r = new SmartRange(0, 10, 2) // 0, 2, 4, 6, 8
+ * r.length // 5
+ * const a = [...r]; // [0,2,4,6,8]
+ * // stepsize defaults to 1 | -1
+ * new SmartRange(2,-3) // 2, 1, 0, -1, -2
+ * ```
+ * @implements {Iterable<number>}
+ * @implements {Iterator<number, undefined>}
+ * @implements {ArrayLike<number>}
+ * @author Leo Schurrer <l.schurrer@outlook.com>
+ * @license MIT
+ */
+export default class SmartRange
+    implements ArrayLike<number>, Iterator<number, undefined>, Iterable<number>
+{
+    // index signature only usefull if class is used with Proxy
+    [index: number]: number;
+
+    // private members
+    #start: number;
+    #end: number;
+    #step: number;
+    /** number of steps performed by the Iterator protocol */
+    #doneSteps: number = 0;
+
+    /**
+     * ## Constructor
+     * values need to be integers
+     * @param start start of the range can be larger than end with negative step
+     * @param end end of the range
+     * @param step step between values in the range - can be negative defaults to 1 or -1 if end < start
+     */
+    constructor(start: number, end: number, step?: number) {
+        //  TODO accept options for excludeStart, includeEnd
```

Comment:
Maybe remove this TODO and add a issue for this topic

Reply:
✔